### DISCUSSION TITLE
are all operations lock free?

### DISCUSSION THREAD

**Discussion by kapitainsky** [rustic-rs/rustic#1316] [Q&A] [answered]
Maybe trivial question but I could not find clarification anywhere...

The story of two laptops backing up to the same cloud remote. I know that there is no issue with running concurrent backups - this is how it works for me for long. The same with one laptop backing up and other running prune.

But what about other operations? Can I run two `rustic prune`? (of course without `--instant-delete` flag) at the same time?. And what about `rustic check`?

In general are all operations lock free and multiple can be run at the same time or there are some exceptions (like `--instant-delete` for prune).

**Comment by aawsome** [comment, marked as answer]
Hi @kapitainsky 
Yes, all operations are designed lock-free. This means all commands can run parallel. This is especially true for multiple backup runs and backup runs parallel to `prune`. 
However, make sure that each individual `backup` run won't take longer as the `--keep-delete` option (default: 23h) if you run `prune` parallel to `backup` - and that you don't use `--instant-delete`.

Of course, any read-only command also safely runs parallel to any other command.

Multiple parallel `forget` or `prune` runs are designed to work, too. But I have to admit that we have not tested this in detail so far. Because of this and other reasons (like ACLs to set up and scheduling) I would recommend to schedule `forget` and `prune` for the whole repository using a single host to do so. However feedback on this with multiple parallel runs is highly appreciated.

There is one general caveat: Error handling is not perfect yet in rustic and we may run into cases where parallel runs may throw errors. E.g. the combination `forget`/`check` may lead to `check` trying to load just-removed snapshots. Or `forget`/`forget` may want to remove a just-removed snapshots which then fails. These are however all cases which should not affect the general repository consistency. If you encounter such an error handling problem, please report!

**Reply by kapitainsky** [reply]
Thank you for clarification.

**Reply by simonsan** [reply]
Added your answer to the FAQ: https://rustic.cli.rs/docs/FAQ.html#are-all-operations-lock-free
